### PR TITLE
INT-3935: Add Jsch Proxy Factory Bean

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3935

### CODE REVIEW

File: spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/JschProxyFactoryBean.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		case SOCKS4:
+			ProxySOCKS4 socks4proxy = new ProxySOCKS4(this.host, this.port);
+			socks4proxy.setUserPasswd(this.user, this.password);
+			return socks4proxy;
+		case HTTP:
+			ProxyHTTP httpProxy = new ProxyHTTP(this.host, this.port);
+			httpProxy.setUserPasswd(this.user, this.password);
+			return httpProxy;
+		default:
+			throw new UnsupportedOperationException("Invalid type:" + this.type);
```
Comment (artembilan) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/JschProxyFactoryBean.java]: This exception doesn't sound good here.
I feel like one of the `FatalBeanException` fits better here.
`InvalidPropertyException`?
`BeanDefinitionStoreException`?

Comment (garyrussell) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/JschProxyFactoryBean.java]: It doesn't really matter - it's just there to satisfy the compiler; it's not actually reachable.

Feel free to change it during merge if you feel strongly.

Comment (artembilan) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/JschProxyFactoryBean.java]: Doh. Of course! We are here only with the `Type` enum! :smile: 

Strange Java which requires from us such a useless, noisy code...

File: src/reference/asciidoc/sftp.adoc [hunk 2]
```diff
@@ -177,6 +178,32 @@ Also see `allowUnknownHosts`.
...
+----
+<bean id="proxySocks5" class="org.springframework.integration.sftp.session.JschProxyFactoryBean">
+    <constructor-arg value="SOCKS5" />
+    <constructor-arg value="${sftp.proxy.address}" />
+    <constructor-arg value="${sftp.proxy.port}" />
+    <constructor-arg value="${sftp.proxy.user}" />
+    <constructor-arg value="${sftp.proxy.pw}" />
+</bean>
+
+<bean id="sessionFactory" class="" >
```
Comment (artembilan) [inline src/reference/asciidoc/sftp.adoc]: Which class?
I think for newbie it would be confusing...

Comment (garyrussell) [inline src/reference/asciidoc/sftp.adoc]: Doh - please add during merge.

**Comment by artembilan** [comment]
Looks like this:
> Host should not be required when a proxy is provided.

Is invalid anyway?
```
public Session getSession(String username, String host, int port) throws JSchException {
    if(host==null){
      throw new JSchException("host must not be null.");
    }
```

**Comment by garyrussell** [comment]
No; that was my mistake - of course you still need a host to connect to __via__ the proxy.

client->proxy->target.

So there are two hosts.

**Comment by artembilan** [comment]
> client->proxy->target.

Yeah! That's correct.
Merging with that simple polishing to the docs...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/62db308c22f44e32eddc15c0b825e91e12caa16b

**Pull request closed without merging**